elyra/pipeline/kfp/kfp_authentication.py

```diff
@@ -0,0 +1,739 @@
+                resp = s.get(redirect_url_obj.geturl(), allow_redirects=True)
+                request_history.append((redirect_url_obj.geturl(), resp))
+                if resp.status_code != HTTPStatus.OK:
+                    raise AuthenticationError('Error redirecting to the DEX login page: '
+                                              f'HTTP status code {resp.status_code}.',
+                                              provider=self._type,
+                                              request_history=request_history)
+                # set the login url
+                dex_login_url = resp.url
+
+            ################
+            # Attempt Dex Login
+            ################
+            resp = s.post(
+                dex_login_url,
+                data={"login": username, "password": password},
+                allow_redirects=True
+            )
+            request_history.append((dex_login_url, resp))
+
+            if len(resp.history) == 0:
+                raise AuthenticationError('The LDAP credentials are probably invalid. '
+                                          f'Update runtime configuration \'{runtime_config_name}\' and try again.',
+                                          provider=self._type,
+                                          request_history=request_history)
+
+            # store the session cookies in a "key1=value1; key2=value2" string
+            return "; ".join([f"{c.name}={c.value}" for c in s.cookies])
+
+        # this code should never be reached; raise an error
+        raise AuthenticationError('An implementation problem was detected for LDAP authentication. '
+                                  'Please create an issue.',
+                                  provider=self._type,
+                                  request_history=request_history)
+
+
+class K8sServiceAccountTokenAuthenticator(AbstractAuthenticator):
+    """
+    Authenticator for Service Account Tokens on Kubernetes.
+    """
+
+    _type = SupportedAuthProviders.KUBERNETES_SERVICEACCOUNT_TOKEN
+
+    def authenticate(self,
+                     kf_endpoint: str,
+                     runtime_config_name: str) -> None:
+        """
+        Verify that service account token authentication can be performed.
+        An AuthenticationError is raised if a problem is encountered that
+        would likely prevent the KFP client from authenticating successfully.
+
+        :param kf_endpoint: Kubeflow API endpoint to verify
+        :type kf_endpoint: str
+        :param runtime_config_name: Runtime configuration name where kf_endpoint is specified
+        :type runtime_config_name: str
+        :raises AuthenticationError: A potential issue was detected that will
+        likely cause a KFP client failure.
+        """
+
+        request_history = []
+
+        # Verify the API endpoint
+        resp = requests.get(kf_endpoint, allow_redirects=True)
```

Comment:
this will continue to fail if we, as users specify an internal endpoint resolved by the kubedns, giving kubeflow as an example, ml-pipeline.kubeflow.svc:8888 doesn't serve an GET request on the base path, neither any other http method.

that will wrongly fail saying that the url is invalid on unreachable and that's not true, when sent to `kfp.Client` it handles it properly.
i don't get why is this check 🤷 

another thing, if this is a secure authenticator, where are you doing any kind of authentication when performing the request? If that endpoint is secure, it is more then acceptable that it returns not found if you don't do any kind of authentication.

And, in a standard way of doing this, the ideal method to do this would be OPTIONS, which is used to describe possible methods that are accepted at the given endpoint.

Reply:
For `K8sServiceAccountTokenAuthenticator` the underlying idea was to determine whether there are any missing prerequisites that would cause the kfp sdk client to fail, not to actually perform authentication. For example, if the resolved service account token path is invalid, we know that the client won't be able to connect.

I don't think adding more options to cover _every possible configuration scenario_ is the way to go. If my assumption was wrong that the provided URL can always be resolved then I'll remove the connectivity test.